main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
     ///
     /// Helper for toList.
     ///
-    def toListHelper(a: Array[a], i: Int): List[a] & Impure =
+    def toListHelper(a: Array[a], i: Int32): List[a] & Impure =
         if (i == length(a))
             Nil
         else
             a[i] :: toListHelper(a, i + 1)
 
-}
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `a`.
+    ///
+    /// Returns `None` if `a` is empty.
+    ///
+    pub def head(a: Array[a]): Option[a] & Pure =
+        if (length(a) > 0) Some(a[0] as & Pure) else None
+
+    ///
+    /// Returns `Some(x)` if `x` is the last element of `a`.
+    ///
+    /// Returns `None` if `a` is empty.
+    ///
+    pub def last(a: Array[a]): Option[a] & Pure =
+        let len = length(a);
+        if (len > 0) Some(a[len-1] as & Pure) else None
+
+    ///
+    /// Return a new array, appending the elements `b` to elements of `a`.
+    ///
+    pub def append(a: Array[a], b: Array[b]): Array[a] & Impure =
+        appendHelper(a, b)
+    
+    /// 
+    /// Helper for `append` (impure).
+    def appendHelper(a: Array[a], b: Array[b]): Array[a] & Impure =
+        let len1 = length(a);
+        let len2 = length(b);
+        if (len1 == 0)
+            slice(b, 0, len2)
+        else {
+            let out = new(a[0], len1 + len2);
+            updateSequenceInPlace(0, a, out);
+            updateSequenceInPlace(len1, b, out);
+            out
+        }
+
+    ///
+    /// Returns `true` if and only if `a` contains the element `x`.
+    ///
+    pub def memberOf(x: a, a: Array[a]): Bool =
+        exists(y -> y == x, a)
+
+    ///
+    /// Alias for `IndexOfLeft`
+    ///
+    pub def indexOf(x: a, a: Array[a]): Option[Int32] = indexOfLeft(x,a)
+
+    ///
+    /// Optionally returns the position of the first occurence of `x` in `a`
+    /// searching from left to right.
+    ///
+    pub def indexOfLeft(x: a, a: Array[a]): Option[Int32] & Pure =
+        let i = indexOfLeftHelper(x, a, length(a), 0) as & Pure;
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Helper function for `indexOfLeft`.
+    ///
+    def indexOfLeftHelper(x: a, a: Array[a], len:Int32, i: Int32): Int32 & Impure =
+        if (i >= len)
+            -1
+        else
+            if (a[i] == x) i else indexOfLeftHelper(x, a, len, i+1)
+
+    ///
+    /// Optionally returns the position of the first occurence of `x` in `a`
+    /// searching from right to left.
+    ///
+    pub def indexOfRight(x: a, a: Array[a]): Option[Int32] & Pure =
+        let start = length(a) - 1;
+        let i = indexOfRightHelper(x, a, start) as & Pure;
+        if (i < 0) None else Some(i)
+
+    ///
+    /// Helper function for `indexOfRight`.
+    ///
+    def indexOfRightHelper(x: a, a: Array[a], i: Int32): Int32 & Impure =
+        if (i < 0)
+            -1
+        else
+            if (a[i] == x) i else indexOfRightHelper(x, a, i-1)
```

Comment:
When I see this code, I wonder if it was a mistake to remove our switch expression:

https://github.com/flix/flix/pull/976
https://flix.dev/#/blog/design-flaws-in-flix/

Do you have any opinion on the matter? If we had some variant of switch back, we could write just:

```scala
def indexOfRightHelper(x: a, a: Array[a], i: Int32): Int32 & Impure = switch {
  case i <0       => -1
  case a[i] == x  => i
  else            => indexOfRightHelper(x, a, i-1)
}
```

This is not really important, but I just wonder, so let me know if you have any opinion.